### PR TITLE
Install native table extension

### PR DESCRIPTION
**Pull request by msmouse**
## Motivation

1. Support table items in the state db.
2. Copy over the Table.move file from upstream
3. Add /tables/<handle>/item API
4. Convert the NFT smoke test

TODO:
Update sdks
Update tutorials

### Have you read the [Contributing Guidelines on pull requests](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md#pull-requests)?

Y

## Test Plan

Existing coverage

## Related PRs

Depends on these to compile and pass tests:
https://github.com/diem/move/pull/181
https://github.com/diem/move/pull/194

### CODE REVIEW

File: aptos-move/aptos-vm/src/aptos_vm_impl.rs [hunk 1]
```diff
@@ -610,7 +618,26 @@ pub fn convert_changeset_and_events_cached<C: AccessPathCache>(
...
+
+pub fn convert_table_changeset(
+    table_changeset: TableChangeSet,
+    out_write_set: &mut WriteSetMut,
+) -> Result<(), VMStatus> {
+    for (handle, change) in table_changeset.changes {
+        for (key, value_opt) in change.entries {
+            let state_key = StateKey::table_item(handle.0, key);
+            if let Some(bytes) = value_opt {
+                out_write_set.push((state_key, WriteOp::Value(bytes)))
```
Comment (msmouse) [inline aptos-move/aptos-vm/src/aptos_vm_impl.rs:633]: @CapCap  So for a table item, the DB doesn't know it's key_type nor the value_type. so basically just opaque bytes (for the handle, the key and the value). Does the indexer want it? (playing with it here: https://github.com/aptos-labs/aptos-core/commit/dd61bc8c97932970c6e1c8c2db39a787ed0faa1b)

File: api/move-test-package/sources/TableTestData.move [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        Table::add(&mut t.u8_table, &1, 1);
+        Table::add(&mut t.u64_table, &1, 1);
+        Table::add(&mut t.u128_table, &1, 1);
+        Table::add(&mut t.bool_table, &true, true);
+        Table::add(&mut t.string_table, &str, copy str);
+        Table::add(&mut t.address_table, &@0x1, @0x1);
+        Table::add(&mut t.vector_u8_table, &vec_u8, copy vec_u8);
+        Table::add(&mut t.vector_string_table, &vec_str, copy vec_str);
+        Table::add(&mut t.id_table, &id, copy id);
+        Table::add(&mut t.table_table, &1, table_u8);
```
Comment (zekun000) [inline api/move-test-package/sources/TableTestData.move:59]: those tables only have one item, probably should test for many items as well?

File: api/src/state.rs [hunk 3]
```diff
@@ -149,4 +179,36 @@ impl State {
...
+        let key_type = key_type.try_into()?;
+        let value_type = value_type.try_into()?;
+
+        let resolver = self.state_view.as_move_resolver();
+        let converter = resolver.as_converter();
+
+        let vm_key = converter
+            .try_into_vm_value(&key_type, key.clone())
+            .map_err(Error::bad_request)?;
+        println!("move value: {:?}", vm_key);
```
Comment (zekun000) [inline api/src/state.rs:198]: remove?

File: api/src/state.rs [hunk 4]
```diff
@@ -149,4 +179,36 @@ impl State {
...
+        let raw_key = vm_key
+            .undecorate()
+            .simple_serialize()
+            .ok_or_else(|| Error::internal(anyhow!("Key failed to serialize.")))?;
+        println!("raw key: {:?}", raw_key);
```
Comment (zekun000) [inline api/src/state.rs:203]: ditto

File: api/src/tests/state_test.rs [hunk 5]
```diff
@@ -85,10 +95,167 @@ async fn test_get_account_module_not_found() {
...
+        install_dir: Some(package_dir.clone()),
+        additional_named_addresses: [("TestAccount".to_string(), account)].into(),
+        ..Default::default()
+    };
+    let _package = build_config
+        .compile_package(&package_dir, &mut Vec::new())
+        .unwrap();
+
+    let mut out = Vec::new();
+    tokio::fs::File::open(package_dir.join("build/ApiTest/bytecode_modules/TableTestData.mv"))
```
Comment (zekun000) [inline api/src/tests/state_test.rs:223]: I don't think you need to read from file, it's in the compiled package. see https://github.com/aptos-labs/aptos-core/blob/b1601d23fb157fee4c4932cd093d3809be8094a4/testsuite/smoke-test/src/aptos/module_publish.rs#L34

File: aptos-move/framework/aptos-framework/sources/Table.move [hunk 6]
```diff
@@ -1,152 +1,108 @@
-/// This module provides a temporary solution for tables by providing a layer on top of Vector
+/// Type of large-scale storage tables.
```
Comment (zekun000) [inline aptos-move/framework/aptos-framework/sources/Table.move:1]: probably need more comments than this lol

**Review by zekun000 — COMMENTED** [review]
amazing! such an exciting feature

File: aptos-move/aptos-vm/src/aptos_vm_impl.rs (hunk 1 above)
Comment (CapCap) [reply aptos-move/aptos-vm/src/aptos_vm_impl.rs:633]: Ouch :-( is there a way to get or store the types for the write set?

Comment (CapCap) [reply aptos-move/aptos-vm/src/aptos_vm_impl.rs:633]: We’d still want it either way, but more grokable is better ;-)

File: aptos-move/framework/aptos-framework/sources/Table.move [hunk 7]
```diff
@@ -1,152 +1,108 @@
...
-        key: &Key,
-    ): &mut Value {
-        let maybe_idx = find(table, key);
-        assert!(Option::is_some(&maybe_idx), Errors::invalid_argument(EKEY_NOT_FOUND));
-        let idx = *Option::borrow(&maybe_idx);
-        &mut Vector::borrow_mut(&mut table.data, idx).value
+    /// Add a new entry to the table. Aborts if an entry for this
+    /// key already exists. The entry itself is not stored in the
+    /// table, and cannot be discovered from it.
+    public fun add<K, V>(table: &mut Table<K, V>, key: &K, val: V) {
```
Comment (sitalkedia) [inline aptos-move/framework/aptos-framework/sources/Table.move:35]: Should this be named `insert` instead for consistency with `Map` APIs ?

File: aptos-move/framework/aptos-framework/sources/Table.move [hunk 8]
```diff
@@ -1,152 +1,108 @@
...
-    public fun insert<Key: store, Value: store>(
-        table: &mut Table<Key, Value>,
-        key: Key,
-        value: Value,
-    ) {
-        let maybe_idx = find(table, &key);
-        assert!(Option::is_none(&maybe_idx), Errors::invalid_argument(EKEY_ALREADY_EXISTS));
-        Vector::push_back(&mut table.data, TableElement { key, value });
+    /// Returns the length of the table, i.e. the number of entries.
+    public fun length<K, V>(table: &Table<K, V>): u64 {
```
Comment (sitalkedia) [inline aptos-move/framework/aptos-framework/sources/Table.move:53]: Same as above. May be rename this API to `len()` instead for consistency with hashmap API?

File: aptos-move/framework/aptos-framework/sources/Table.move [hunk 9]
```diff
@@ -1,152 +1,108 @@
...
-    ): &mut Value {
-        let maybe_idx = find(table, key);
-        assert!(Option::is_some(&maybe_idx), Errors::invalid_argument(EKEY_NOT_FOUND));
-        let idx = *Option::borrow(&maybe_idx);
-        &mut Vector::borrow_mut(&mut table.data, idx).value
+    /// Add a new entry to the table. Aborts if an entry for this
+    /// key already exists. The entry itself is not stored in the
+    /// table, and cannot be discovered from it.
+    public fun add<K, V>(table: &mut Table<K, V>, key: &K, val: V) {
+        add_box<K, V, Box<V>>(table, key, Box{val});
```
Comment (sitalkedia) [inline aptos-move/framework/aptos-framework/sources/Table.move:36]: What's the behavior if the key already exists?

File: crates/aptos-rest-client/src/lib.rs [hunk 10]
```diff
@@ -268,27 +266,6 @@ impl Client {
         self.json(response).await
     }
 
-    pub async fn get_account_resources_by_type(
```
Comment (sitalkedia) [inline crates/aptos-rest-client/src/lib.rs:271]: Why is this removed?

File: aptos-move/framework/aptos-framework/sources/Table.move [hunk 11]
```diff
@@ -1,152 +1,108 @@
...
-        Vector::length(&table.data)
-    }
-
-    public fun create<Key: store, Value: store>(): Table<Key, Value> {
-        Table {
-            data: Vector::empty(),
+    /// Create a new Table.
+    public fun new<K, V: store>(): Table<K, V> {
+        Table{
+            handle: new_table_handle(),
```
Comment (sitalkedia) [inline aptos-move/framework/aptos-framework/sources/Table.move:20]: How do we ensure the uniqueness of this handle across tables?

**Review by sitalkedia — CHANGES_REQUESTED** [review]

File: aptos-move/framework/aptos-framework/sources/Table.move [hunk 12]
```diff
@@ -1,152 +1,108 @@
...
+    public fun borrow_mut_with_default<K, V: drop>(table: &mut Table<K, V>, key: &K, default: V): &mut V {
+        if (!contains(table, key)) {
+            add(table, key, default)
         };
-
-        Option::none()
+        borrow_mut(table, key)
     }
 
-    #[test]
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/Table.move:98]: why you hate tests? they should still work, right?

File: aptos-move/aptos-vm/src/aptos_vm_impl.rs (hunk 1 above)
Comment (msmouse) [reply aptos-move/aptos-vm/src/aptos_vm_impl.rs:633]: @CapCap  The indexer can in theory discover the type info when the table is first created (as a struct under an resource, because you'll be able to parse the type info out of the resource embodying resource type.), but obviously that's not too trivial.

@sitalkedia  ^ the type info is in the table itself, as all other parts of the system: typing info is in the modules,  separate from the instance data.

File: api/move-test-package/sources/TableTestData.move (hunk 2 above)
Comment (msmouse) [reply api/move-test-package/sources/TableTestData.move:59]: To the storage it's all the same. (Sizing related stuff is tested in the module itself.)

File: api/src/tests/state_test.rs (hunk 5 above)
Comment (msmouse) [reply api/src/tests/state_test.rs:223]: 👌 (But we'll serialize it again!)

File: api/src/state.rs [hunk 13]
```diff
@@ -149,4 +179,36 @@ impl State {
...
+
+        let state_key = StateKey::table_item(handle, raw_key);
+        let bytes = self
+            .state_view
+            .get_state_value(&state_key)?
+            .ok_or_else(|| Error::not_found("table item", key, self.ledger_version))?;
```
[the remark this answers is not in the document]
Comment (msmouse) [reply api/src/state.rs:209]: We can't because we don't know where the Table struct is defined, and we also don't really want to do the extra read even if we know.

File: api/move-test-package/sources/TableTestData.move [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+    struct TestTables has key {
+        u8_table: Table<u8, u8>,
+        u64_table: Table<u64, u64>,
+        u128_table: Table<u128, u128>,
+        bool_table: Table<bool, bool>,
+        string_table: Table<ASCII::String, ASCII::String>,
+        address_table: Table<address, address>,
+        vector_u8_table: Table<vector<u8>, vector<u8>>,
+        vector_string_table: Table<vector<ASCII::String>, vector<ASCII::String>>,
```
[the remark this answers is not in the document]
Comment (msmouse) [reply api/move-test-package/sources/TableTestData.move:16]: The key and value types don't relate in any way in the code so these should suffice?

File: aptos-move/aptos-vm/src/parallel_executor/read_write_set_analyzer.rs [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    }
+}
+
+impl<'a, S: MoveResolverExt + std::marker::Sync> ReadWriteSetInferencer
+    for ReadWriteSetAnalysisWrapper<'a, S>
+{
+    type T = PreprocessedTransaction;
+    fn infer_reads_writes(&self, txn: &Self::T) -> Result<Accesses<StateKey>> {
+        let (keys_read, keys_written) = self.analyzer.get_keys_transaction(txn, false)?;
+        // TODO: Add support for table items as state key.
```
[the remark this answers is not in the document]
Comment (msmouse) [reply aptos-move/aptos-vm/src/parallel_executor/read_write_set_analyzer.rs:31]: looks like the whole file is removed, and this is a rebase error.

[the remark this answers is not in the document]
Comment (msmouse) [reply aptos-move/aptos-vm/src/parallel_executor/read_write_set_analyzer.rs:31]: Looks like the whole file is gone, and this is a rebase error. (Because the new Parallel Executor doesn't rely on predictions)

File: aptos-move/framework/aptos-framework/sources/Table.move (hunk 7 above)
Comment (msmouse) [reply aptos-move/framework/aptos-framework/sources/Table.move:35]: This is a upstream file copied over. And the upstream deliberately tries to distinguish the Table with a Map IIUC 😂

File: aptos-move/framework/aptos-framework/sources/Table.move (hunk 9 above)
Comment (msmouse) [reply aptos-move/framework/aptos-framework/sources/Table.move:36]: It's an error. https://github.com/move-language/move/blob/be45aa0cc24f8b7106979bdf3313c70eb8d16a12/language/extensions/move-table-extension/src/lib.rs#L222-L226

File: aptos-move/framework/aptos-framework/sources/Table.move (hunk 11 above)
Comment (msmouse) [reply aptos-move/framework/aptos-framework/sources/Table.move:20]: This is the implementation: https://github.com/move-language/move/blob/be45aa0cc24f8b7106979bdf3313c70eb8d16a12/language/extensions/move-table-extension/src/lib.rs#L336-L362

Handles are based on the txn id and the sequence number among the tables being created in the same table.

File: aptos-move/framework/aptos-framework/sources/Table.move (hunk 12 above)
Comment (msmouse) [reply aptos-move/framework/aptos-framework/sources/Table.move:98]: This file is a copy from upstream, exact same file, and it's tested in a separate unit test file upstream.

File: aptos-move/aptos-vm/src/move_vm_ext/vm.rs [hunk 16]
```diff
@@ -21,13 +21,13 @@ impl MoveVmExt {
...
+    pub fn new_session<'r, S: MoveResolverExt>(
         &self,
         remote: &'r S,
-        _session_id: SessionId,
+        session_id: SessionId,
     ) -> SessionExt<'r, '_, S> {
-        // TODO: install table extension
-        let extensions = NativeContextExtensions::<'r>::default();
+        let mut extensions = NativeContextExtensions::default();
+        extensions.add(NativeTableContext::new(session_id.as_uuid(), remote));
```
Comment (zekun000) [inline aptos-move/aptos-vm/src/move_vm_ext/vm.rs:30]: cc: @gelash here's an example how we pass a context to native functions

File: aptos-move/framework/aptos-framework/sources/Table.move (hunk 8 above)
Comment (msmouse) [reply aptos-move/framework/aptos-framework/sources/Table.move:53]: `length()` is consistent with `Vector::length()` (in MOVE)

File: aptos-move/aptos-vm/src/parallel_executor/read_write_set_analyzer.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (zekun000) [reply aptos-move/aptos-vm/src/parallel_executor/read_write_set_analyzer.rs:31]: remove the file?

File: aptos-move/framework/aptos-framework/sources/Table.move [hunk 17]
```diff
@@ -1,152 +1,113 @@
...
+    const EALREADY_EXISTS: u64 = 100;
+    // native code raises this with Errors::invalid_arguments()
+    const ENOT_FOUND: u64 = 101;
+    const ENOT_EMPTY: u64 = 102;
 
-    struct Table<Key: store, Value: store> has store {
-      data: vector<TableElement<Key, Value>>,
+    /// Type of tables
+    struct Table<phantom K, phantom V> has store {
+        handle: u128,
```
Comment (zekun000) [inline aptos-move/framework/aptos-framework/sources/Table.move:18]: why we use u128 instead of u64?

File: crates/aptos-workspace-hack/Cargo.toml [hunk 18]
```diff
@@ -41,6 +41,8 @@ itertools = { version = "0.10.3", features = ["use_alloc", "use_std"] }
 libc = { version = "0.2.123", features = ["std"] }
 log = { version = "0.4.16", default-features = false, features = ["serde", "std"] }
 memchr = { version = "2.4.1", features = ["std", "use_std"] }
+move-unit-test = { git = "https://github.com/move-language/move", rev = "476305d239ce6afafce15a297c8c3839dd8465b6", default-features = false, features = ["table-extension"] }
+move-vm-test-utils = { git = "https://github.com/move-language/move", rev = "476305d239ce6afafce15a297c8c3839dd8465b6", features = ["move-table-extension", "table-extension"] }
```
Comment (zekun000) [inline crates/aptos-workspace-hack/Cargo.toml:45]: why there's two features?

**Review by zekun000 — APPROVED** [review]

File: api/src/state.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (sitalkedia) [reply api/src/state.rs:207]: If we don't want to do an extra read for the table metadata here, let's improve the error message saying "table or table item not found" ?

**Review by sitalkedia — APPROVED** [review]

File: aptos-move/aptos-vm/src/aptos_vm_impl.rs (hunk 1 above)
Comment (CapCap) [reply aptos-move/aptos-vm/src/aptos_vm_impl.rs:618]: that actually sounds pretty good :-) can have a PG table for those table types :-)

File: aptos-move/framework/aptos-framework/sources/Table.move (hunk 17 above)
Comment (msmouse) [reply aptos-move/framework/aptos-framework/sources/Table.move:18]: We actually want to move to U256, especially if you ask @alinush

File: crates/aptos-workspace-hack/Cargo.toml (hunk 18 above)
Comment (msmouse) [reply crates/aptos-workspace-hack/Cargo.toml:45]: There isn't really a "move-table-extension" feature.. I think it's a bug in x..

File: aptos-move/aptos-vm/src/aptos_vm_impl.rs [hunk 19]
```diff
@@ -549,17 +550,28 @@ pub fn convert_changeset_and_events(
     changeset: MoveChangeSet,
     events: Vec<MoveEvent>,
 ) -> Result<(WriteSet, Vec<ContractEvent>), VMStatus> {
-    convert_changeset_and_events_cached(&mut (), changeset, events)
+    let mut out_write_set = WriteSetMut::new(vec![]);
```
[the remark this answers is not in the document]
Comment (msmouse) [reply aptos-move/aptos-vm/src/aptos_vm_impl.rs:553]: e2e-test-replay and encode_genesis_change_set, I might need to fix them down the road..

**Comment by msmouse** [comment]
> If we don't want to do an extra read for the table metadata here, let's improve the error message saying "table or table item not found" ?
@sitalkedia  did that but notice, there's not a table metadata item
